### PR TITLE
[RNMobile] Update Inserter Block Search Styling

### PR DESCRIPTION
**Pull request by jhnstn**
## Description
This updates the inserter block search styling to handle variances between iOS and Android. It also introduces a no results view.

## How has this been tested?
1. Load the demo app in both IOS and Android
2. Open the Inserter
3. Verify that the styling looks correct in light and dark mode
4. Press on the search input
5. Verify that the styling for active search looks correct.
6. Try entering any text.
7. Verify that the cancel input options look correct and clear out the text input when pressed
8. Enter more text and press on the cancel search options.
9. Verify that the bottom sheet is closed without inserting a block
10. Open the Inserter and enter arbitrary text that does not match to a block e.g. `asdf`
11.  Verify that the "No block found" message is displayed.
12. Press backspace on the inputed text 
13. Verify that possible query results are rendered in the inserter e.g. removing characters from `asdf` until only `as` remains.

### Open Inserter

Android|iOS
--|--
Screenshot_20210708-155752|IMG_0402
Screenshot_20210708-155916|IMG_0407

### Start Search
Android|iOS
--|--
Screenshot_20210708-162619|IMG_0403
Screenshot_20210708-165018|IMG_0405

### No results

Android|iOS
--|--
Screenshot_20210708-162641|IMG_0404
Screenshot_20210708-162532|IMG_0406

## Types of changes

This adds new mobile platform specific style sheets and a new native component to show the no results view.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [inline packages/components/src/search-control/index.native.js:91]: The idea here is to use a  hook to build the styles on demand given any state changes. It avoids having to weave conditionals in the components per state e.g. `<someComponent style = { isActive ? activeStyle : style } />` . This approach can also be used to replace the `usePreferredColorSchemeStyle` which needs to be called on every individual set of style rules. Instead one single call can be made on the entire style object to generate the themed styles.

I'd like to extract this way of handling the styles out to a hook since it's a more declarative way to build the style object for the component. In the interest of time however, I opted to leave this idea scoped to this module.

The signature of the hook would looks something like 
```
const currentStyles = useModifiedStyles(allStyles, 
  [ 
    {'modifierA' : [someState]} , 
    {'modifierB': [someState, someOtherState]}
  ]
)
```

**Comment by fluiddot** [comment]
I'm concerned about the potential side effects of not having the segmented control along with the search (this only happens in dev mode), I remember when I was working on the reusable block that I experienced some issues 🤔 .

Screenshot 2021-07-09 at 16 55 30

I have a fix for this issue in [this PR](https://github.com/WordPress/gutenberg/pull/32985) but it's not approved yet, it would be useful to use it to test this PR as it will help to test in a similar environment as production.

**Comment by fluiddot** [comment]
👋 Thanks @jhnstn for updating the styles of the search form, now it looks awesome!

After testing the search functionality on both iOS and Android, I noticed a the following issue:

### 1. Search result list requires some padding at the bottom for the safe area [iOS][medium]

I noticed this in the search result list, however, it's not happening on the regular list.

_Reproduced on iPhone 8 - simulator (iOS 14.2) | Medium impact_



### 2. Search input field can be scrolled some pixels

_Reproduced on Samsung Galaxy S20 FE 5G (Android 10) | Low impact_
 
https://user-images.githubusercontent.com/14905380/125100658-a4a90f00-e0d9-11eb-8f55-c216523f358f.mp4

### 3. Some items disappear when searching

I reproduced this issue by using specific letters in the search, in the example below it's the letter `i`, but I think it could happen with others too. I don't mean that is related to the search term but probably to some layout animation that is triggered when filtering the items after searching 🤔 .

_Reproduced on Samsung Galaxy S20 FE 5G (Android 10) | High impact_

https://user-images.githubusercontent.com/14905380/125100962-f6ea3000-e0d9-11eb-9cf4-80f656679b17.mp4

### 4. Debounce search

This is just more a recommendation but totally optional, I was wondering if we would benefit from using debounce when typing the search term, I guess the logic for searching is fast enough to maybe not require it, wdyt?

File: packages/block-editor/src/components/inserter/search-results.native.js
Comment (fluiddot) [inline packages/block-editor/src/components/inserter/search-results.native.js]: Probably an edge case but I was wondering if we should render the no results component in case `items` is undefined, wdyt?

File: packages/block-editor/src/components/inserter/style.native.scss
Comment (fluiddot) [inline packages/block-editor/src/components/inserter/style.native.scss]: I think we don't need to include the `font-size` here as it will take the value from the base variant (light) in line 35.

File: packages/components/src/search-control/searchFormStyles.ios.scss
Comment (fluiddot) [inline packages/components/src/search-control/searchFormStyles.ios.scss]: I can't find where we're using the styles defined in this file, are we really using them?

File: packages/components/src/search-control/searchFormStyles.android.scss
Comment (fluiddot) [inline packages/components/src/search-control/searchFormStyles.android.scss]: I can't find where we're using the styles defined in this file, are we really using them?

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [inline packages/components/src/search-control/index.native.js]: Could we include `alignSelf` to the `cancelButtonStyle` style?

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [inline packages/components/src/search-control/index.native.js]: I'm wondering if we could have these inline styles in the styles file, wdyt?

**Review by fluiddot — COMMENTED** [review]
Here goes my first review pass on the code, I'd like to check deeper the logic for merging the styles in the search control so next week I'll try a second pass, amazing job @jhnstn 🎊 !

File: packages/components/src/search-control/searchFormStyles.ios.scss
Comment (jhnstn) [reply packages/components/src/search-control/searchFormStyles.ios.scss]: Ah, looks like this got erroneously included while I was resolving the `search-control` changes.

File: packages/components/src/search-control/searchFormStyles.android.scss
Comment (jhnstn) [reply packages/components/src/search-control/searchFormStyles.android.scss]: yep, looks like they got added by accident, [see here](https://github.com/WordPress/gutenberg/pull/33237#discussion_r667060154)

File: packages/block-editor/src/components/inserter/style.native.scss
Comment (jhnstn) [reply packages/block-editor/src/components/inserter/style.native.scss]: yep, looks like this got left in there.

**Comment by jhnstn** [comment]
@fluiddot  thanks for the awesome review! 

I'd like to open up separate issues for the first 3 issues you found when testing (nice catches btw). The feature is still behind a dev flag and I feel this PR is already getting big. Thoughts?

On point 4. I thought about that when I started this project but the list of blocks is static and not very large. I figure if we need to fetch blocks over the wire or if the list of blocks gets really long it would be useful to buffer the input. For now it seems like an over optimization to do that. Open to have my mind changed on that however.

I cleaned up the CSS issues you noted. 

Regarding the logic that sets up the styles, I went ahead and extracted that code to a stand alone hook. I figure it will be easier to grok the code since it's now generalized. Overall it's the same approach as in this PR , just abstracted out and (hopefully) simplified a bit. 
The WIP PR is here: #33339 
I branched off this branch but here is the important commit: https://github.com/WordPress/gutenberg/pull/33339/commits/5004d705ace202ce55c4ca652ae7d21085672cd1

I can cherry pick that commit into a fresh branch if we think it's worth adding as a stand alone hook. Merge that then update this branch.

Thanks again for your thoughtful review! 💯

**Comment by fluiddot** [comment]
> I'd like to open up separate issues for the first 3 issues you found when testing (nice catches btw). The feature is still behind a dev flag and I feel this PR is already getting big. Thoughts?

Make sense, let's address them in new PRs 👍 .

> On point 4. I thought about that when I started this project but the list of blocks is static and not very large. I figure if we need to fetch blocks over the wire or if the list of blocks gets really long it would be useful to buffer the input. For now it seems like an over optimization to do that. Open to have my mind changed on that however.

I thought the same, I was just wondering if we could benefit from it but I agree that it's an over-optimization as calculating the search result should be quite fast.

> I cleaned up the CSS issues you noted.

💯 

> Regarding the logic that sets up the styles, I went ahead and extracted that code to a stand alone hook. I figure it will be easier to grok the code since it's now generalized. Overall it's the same approach as in this PR , just abstracted out and (hopefully) simplified a bit.
> The WIP PR is here: #33339
> I branched off this branch but here is the important commit: [5004d70](https://github.com/WordPress/gutenberg/commit/5004d705ace202ce55c4ca652ae7d21085672cd1)
> 
> I can cherry pick that commit into a fresh branch if we think it's worth adding as a stand alone hook. Merge that then update this branch.

To be honest, I don't think it's necessary if we plan to update this part after merging the stand-alone hook you mentioned.

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [inline packages/components/src/search-control/index.native.js]: `mergeFutureStyles` function only has two arguments so I guess we could remove the third one passed here with value `true`, right?

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [inline packages/components/src/search-control/index.native.js]: I'm a bit concerned about modifying the imported `baseStyles` because it will modify the styles for the rest of the components that might import them, what do you think about creating a new constant?

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [reply packages/components/src/search-control/index.native.js]: yep, That got left in after debugging.

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [reply packages/components/src/search-control/index.native.js]: Good catch. Changing now.

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [inline packages/components/src/search-control/index.native.js:33]: @fluiddot  I just went with a spread here which I know will only do a one level deep copy. I'm not sure if our style setup allows sass type nesting (now or in the future). My feeling is that this should be safe enough, thoughts?

Comment (fluiddot) [reply packages/components/src/search-control/index.native.js:33]: The StyleSheet object of React native only admits one level so it should be safe 👍 .

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [inline packages/components/src/search-control/index.native.js:81]: If this is a static object that won't change, I'm wondering if it would make sense to have it as constant outside the functional component, as we have the base styles, wdyt?

The same comments apply to the rest of the styles:
- darkStyles
- activeStyles
- activeDarkStyles

Comment (jhnstn) [reply packages/components/src/search-control/index.native.js:81]: I noticed that but left these here to show the evolution on extracting the pattern to a hook. Here is the abstraction around how the modified styles are selected from the base styles: https://github.com/WordPress/gutenberg/commit/5004d705ace202ce55c4ca652ae7d21085672cd1#diff-03ee4c0d07daa51fb0c8182f80cdd50c7f5317727593a6a75ea3f877c744f610R96 

I don't mind moving these outside of the component but figure `useMemo` should give us enough cacheing to be performant.

Comment (fluiddot) [reply packages/components/src/search-control/index.native.js:81]: Ok I see, then we can leave them here 👍, thanks for the clarification 🙇‍♂️.

**Comment by guarani** [comment]
Sorry for the delays getting around to this, @jhnstn! I've tested this today and here are some comments.

## Android testing

Tested on Android emulator Nexus 6 API 29. 

### Placeholder disappears when cancelling active search

On Android, after typing text into the search field and then tapping the arrow button, the keyboard is closed (as expected) but the placeholder "Search Blocks" doesn't get shown in the empty search field. Subsequently tapping on the search field doesn't bring up the keyboard and I can't continue searching until I close the bottom sheet.

https://user-images.githubusercontent.com/1898325/125540180-580de082-24c1-40e5-9ff2-618402f9c2ad.mov

### The search field bounces up and is slow to come back down

After opening the block inserter, I tap the search field to start filtering blocks and notice the animation slides the search field up momentarily before bringing it back down on-screen and the animation seems slow. This was seen on an emulator and also on a device. That said this seems low priority.

https://user-images.githubusercontent.com/1898325/125540487-5a3d9ada-c2f5-40eb-8a62-0e24927e2804.mov

### Space between the bottom sheet drag handle and search field too tight

This is subjective and might be more a design question rather than an implementation question, but my first impression is that this spacing is too tight:

Screen Shot 2021-07-13 at 19 19 17

### Placeholder case

The current placeholder text "Search Blocks" stands out to me because I would have expected it to be sentence case: "Search blocks"

### Search field doesn't respond to changes in device font size

After changing the device font size to largest, the search field doesn't adapt its height to the larger placeholder font size (text is cut-off). Tested on Samsung S10 physical device with Android 10, font size changed in device Settings app.

 

## iOS testing

I tested on a physical iPhone 11 running iOS 14.5 and quick test on an iPad simulator and things worked great. I only saw the "Space between the bottom sheet drag handle and search field too tight" and the "Placeholder case" issues. 

---

In my opinion, the only thing that seems high priority is the first issue mentioned on Android where the search becomes unusable after cancelling an active search using the left arrow button. Overall it's working great and I've been using the block inserter search a lot lately when using the editor in development.

**Comment by fluiddot** [comment]
> **The search field bounces up and is slow to come back down**
> After opening the block inserter, I tap the search field to start filtering blocks and notice the animation slides the search field up momentarily before bringing it back down on-screen and the animation seems slow. This was seen on an emulator and also on a device. That said this seems low priority.

This behavior is on purpose because on Android, we don't get notified when the opening transition of the keyboard starts to calculate the final height of the bottom sheet. The workaround I applied for this issue was to make the transition you mentioned that happens when the keyboard transition finishes.

There're further details in the following issue: https://github.com/WordPress/gutenberg/issues/31151

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [inline packages/components/src/search-control/index.native.js:146]: @guarani  So this is an odd one. In the dev tools debugger, I could see the `<TextInput />` in the component tree after canceling the search and there is nothing in the code that would hide or remove that component. 
I had a feeling there might be an animation issue ( somewhat related to the missing items in the `<FlatList />` search results component )
Well I added a timeout to see if queuing the onCancel calls would do anything and it seems to work, even with a 0 millisecond delay 😄 

I could reproduce the original issue on my Pixel 4 and it's fixed with the timeout there.

**Comment by jhnstn** [comment]
Thanks @guarani  for the review!

> Search field doesn't respond to changes in device font size

Fixed in 0ae5d0d63cc74c13612815ae39e5a292afb76b45 ( also fixed the input scrolling issue @fluiddot found)

> Placeholder disappears when cancelling active search

Fixed in cac6bf15abb1baa24e5cc1cfd12510bad25d2102  ( that was an odd one, see my inline notes on the commit )

I really don't have a strong opinion on the bottom sheet upper padding or the Sentence vs Title case for the placeholder. Both are valid points but I'll have to differ to @iamthomasbishop to decide if we should change those.

File: packages/components/src/search-control/index.native.js
Comment (fluiddot) [reply packages/components/src/search-control/index.native.js:146]: On RN we have a timer function named `setImmediate` ([reference](https://reactnative.dev/docs/timers)) that executes the code at the end of the current JavaScript execution block, I think we could use it instead of the `setTimeout` function with a zero delay, wdyt?

**Review by fluiddot — APPROVED** [review]
I've done another test on Android (on iOS everyting worked nice 🎊 ) and here are the results:

### Placeholder disappears when cancelling active search

**✅ Fixed**

I tested on my Android device (Samsung Galaxy S20 FE 5G - Android 10) and I couldn't reproduce it.

### The search field bounces up and is slow to come back down

**🟡 Reproduced but it shouldn't block the PR**

As I mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/33237#issuecomment-879680369), this is a known issue that will be addressed in the future.

### Space between the bottom sheet drag handle and search field too tight

**❓ Wait for design feedback input**

I also have the impression that this spacing is too tight. In fact, I think that's the only area of the bottom sheet where the user can swipe down to close it so maybe we should preserve the padding we had previously. In any case, I don't think this issue should block the PR.

### Placeholder case

**🟡 Reproduced but it shouldn't block the PR**

I also expected the placeholder text to be `Search block` instead of `Search Blocks`. I saw that this change was introduced when updating this branch with `trunk` ([commit reference](https://github.com/WordPress/gutenberg/commit/de815f4bcfbb4bdbc5f7c0bd4bc5c64c90199ae3#diff-583bc205ce379f3d70e7af7aa4d41a21f07925b87d993fc114e4ac4e2f98e67d)), what's the purpose of changing this string?

### Search field doesn't respond to changes in device font size

**✅ Fixed**

**✅ I verified that the issue I [mentioned in a previous comment](https://github.com/WordPress/gutenberg/pull/33237#issuecomment-877270922) ("Search input field can be scrolled some pixels") was also fixed.**

I tested with the biggest font size on my Android device (Samsung Galaxy S20 FE 5G - Android 10) and the cut-off is addressed.

<hr>

Apart from the tests, I added a comment in the code with a suggestion, it's just a minor change so feel free to skip it.

From my side the PR LGTM 🎊 but it would be nice to have the approval from @guarani too, thanks 🙇 !

**Comment by jhnstn** [comment]
> I also expected the placeholder text to be Search block instead of Search Blocks
Thank @fluiddot for finding that commit. Not sure why that change came in on the merge. Also thanks @guarani for originally pointing it out. I went back to the comps and the placeholder is in sentence case. 
I had remembered incorrectly. 

I also went ahead and changed the case for the other labels and hints in the component.

File: packages/components/src/search-control/index.native.js
Comment (jhnstn) [reply packages/components/src/search-control/index.native.js:146]: Oh that's cool, I'll give it a try. I figured since the 0 delay worked all we need to do is run this closer to the end of the event loop. Sounds like `setImmediate` will do more or less that.

Comment (jhnstn) [reply packages/components/src/search-control/index.native.js:146]: Well that didn't work :( . Not sure why but replacing the timeout with `setImmediate`  is causing the inserter bottom sheet to close and then crashes the demo app. I don't think it's worth digging into it right now given setTimeout works (even though it's not the cleanest).

Thanks for pointing out that reference doc.

**Review by guarani — APPROVED** [review]
I re-tested Android and iOS and don't see any of the issues I noted earlier in https://github.com/WordPress/gutenberg/pull/33237#issuecomment-879514250, except those that @fluiddot noted [here](https://github.com/WordPress/gutenberg/pull/33237#pullrequestreview-707609418) (which I understand are not meant to be addressed here).

File: packages/components/src/search-control/index.native.js
Comment (guarani) [reply packages/components/src/search-control/index.native.js:146]: I see, it does seem like a strange bug. I'm glad the timeout workaround worked though.
TIL about the `setImmediate` – I'm not clear what it is meant for though. It's interesting how there's a `clearImmediate` too. Reading the link there, I guess that's because you could clear the `setImmediate` before the end of the current "JavaScript block". It looks like it's a [non-standard web API](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate), but it might be more standard in RN.

Comment (fluiddot) [reply packages/components/src/search-control/index.native.js:146]: > Well that didn't work :( . Not sure why but replacing the timeout with setImmediate is causing the inserter bottom sheet to close and then crashes the demo app. I don't think it's worth digging into it right now given setTimeout works (even though it's not the cleanest).

Oh, that's weird, I tried `setImmediate` locally and in my case it worked but I didn't test it on the demo project, only in the WordPress one 🤔 , how did you use it?

> It's interesting how there's a clearImmediate too. Reading the link there, I guess that's because you could clear the setImmediate before the end of the current "JavaScript block". It looks like it's a non-standard web API, but it might be more standard in RN.

Yeah exactly. To be honest, I haven't used it too often but for some cases, it could be useful as a replacement of `setTimeout`.

Comment (jhnstn) [reply packages/components/src/search-control/index.native.js:146]: > Oh, that's weird, I tried setImmediate locally and in my case it worked but I didn't test it on the demo project, only in the WordPress one 🤔 , how did you use it?

That is odd, I just tried again on the demo app and it worked fine. 

> It looks like it's a non-standard web API, but it might be more standard in RN.

This is first time I've run across this method. In fact it doesn't look like it's part of any modern browser. I don't see it in Chrome or Firefox. 

Which brings up a consideration. It looks like it's well supported in RN but since we are heavily incorporated into a web project and `setTimeout(fn,0)` is a common web pattern I wonder if it's best to stick with the timeout. Thoughts?

Comment (fluiddot) [reply packages/components/src/search-control/index.native.js:146]: I agree that it's not a likely common web pattern but I don't have a strong opinion whether we should use it or not, as a side note, I already used it in the reusable block 😄 :

https://github.com/WordPress/gutenberg/blob/e7df86ec8589cd48ca54e41d37433126094c8777/packages/block-library/src/block/edit.native.js#L140

**Pull request merged**
